File: Assignments/A06/2.py
MOD = 10**9 + 7

def dfs(dp, n, l):
    if dp[l][n]:
        return dp[l][n]
    if l == 1 or n == 1:
        return n
    ret = 0
    for i in range(1, n + 1):
        ret += dfs(dp, i, l - 1) % MOD
        ret %= MOD
    dp[l][n] = ret
    return ret

n, m = map(int, input().split())
dp = [[0]*(n+1) for _ in range(2*m+1)]
print(dfs(dp, n, 2 * m))

File: Assignments/A06/2.cpp
#include <iostream>
#include <vector>
using namespace std;

const int MOD = 1e9 + 7;

int dfs(vector<vector<int>>& dp, int n, int l) {
    if (dp[l][n] != 0) {
        return dp[l][n];
    }
    if (l == 1 || n == 1) {
        return n;
    }
    int ret = 0;
    for (int i = 1; i <= n; i++) {
        ret += dfs(dp, i, l - 1);
        ret %= MOD;
    }
    dp[l][n] = ret;
    return ret;
}

int main() {
    int n, m;
    cin >> n >> m;
    vector<vector<int>> dp(2 * m + 1, vector<int>(n + 1, 0));
    cout << dfs(dp, n, 2 * m) << endl;
    return 0;
}
